File: shipping-service/src/infra/service/message_broker_consumer_service.rs
use crate::{
    application::{
        interfaces::i_message_broker_consumer_service::IMessageBrokerConsumerService,
        usecase::order_accepted_usecase::OrderAcceptedUsecase,
    },
    domain::{enums::e_order_event::EOrderEvent, event::order_event::OrderEvent},
    infra::message_broker::kafka_config::KafkaConsumer,
};

pub struct MessageBrokerConsumerService<'a> {
    kafka_consumer: &'a mut KafkaConsumer,
    order_accepted_use_case: &'a mut OrderAcceptedUsecase<'a>,
}

impl<'a> MessageBrokerConsumerService<'a> {
    pub fn new(
        kafka_consumer: &'a mut KafkaConsumer,
        order_accepted_use_case: &'a mut OrderAcceptedUsecase<'a>,
    ) -> Self {
        Self {
            kafka_consumer,
            order_accepted_use_case,
        }
    }

    fn order_accepted(&mut self, order_event: &OrderEvent) {
        self.order_accepted_use_case.execute(order_event.order_id);
    }
}

impl<'a> IMessageBrokerConsumerService for MessageBrokerConsumerService<'a> {
    fn start_consuming(&mut self) {
        let order_event = &mut OrderEvent {
            event_type: "".to_string(),
            order_id: 0,
        };

        loop {
            let messages = self.kafka_consumer.consume_from_kafka();
            if messages.is_empty() {
                continue;
            }

            for message in messages {
                let order_event_result: Result<OrderEvent, serde_json::Error> =
                    serde_json::from_str(&message);

                match order_event_result {
                    Ok(ok_order_event) => {
                        *order_event = ok_order_event;
                    }
                    Err(e) => {
                        println!("Error consuming message: {}", e);
                        continue;
                    }
                }

                if !order_event.is_valid_to_consume() {
                    continue;
                }

                if order_event.event_type == EOrderEvent::OrderAccepted.to_string() {
                    self.order_accepted(order_event);
                }
            }
        }
    }
}

File: payment-service/src/infra/service/message_broker_consumer_service.py
from application.usecase.attempt_transaction_usecase import AttemptTransactionUseCase
from domain.enums.e_order_event_type import EOrderEventType
from domain.event.order_event import OrderEvent
from infra.message_broker.kafka_config import KafkaConsumer
import json


class MessageBrokerConsumerService:
    def __init__(
        self,
        kafka_consumer: KafkaConsumer,
        attempt_transaction_usecase: AttemptTransactionUseCase,
    ):
        self.kafka_consumer = kafka_consumer
        self.attempt_transaction_usecase = attempt_transaction_usecase

    def start_consuming(self):
        while True:
            message = self.kafka_consumer.consume_from_kafka()
            if message is not None:
                try:
                    self.orderEvent = OrderEvent(**json.loads(message))
                except Exception:
                    continue

                if not self.orderEvent.isValidToConsume():
                    continue

                match self.orderEvent.event_type:
                    case EOrderEventType.INVENTORY_RESERVED.name:
                        self.on_inventory_reserved()
                    case _:
                        continue

        self.kafka_consumer.stop_consuming()

    def on_inventory_reserved(self):
        if not type(self.orderEvent.order_id) is int:
            return
        self.attempt_transaction_usecase.execute(self.orderEvent.order_id)
